### PR TITLE
Allow user to define periodic event type

### PR DESCRIPTION
**Pull request by azadon**
Extending add_periodic with optional argument for configuring event behavior after creation - SD_EVENT_ON, SD_EVENT_OFF or SD_EVENT_ONESHOT. This allows user to create a periodic event which can stop itself from event's callback function.

Withing PR I also added function for getting information if event is enabled or not and description contains also this information.

### CODE REVIEW

File: src/evl.c [hunk 1]
```diff
@@ -10,11 +10,18 @@
...
 	sd_event_source *evsrc = *((sd_event_source**) luaL_checkudata(L, -1, EVSRC_MT));
+
+
+	ret = sd_event_source_get_enabled(evsrc, &enabled);
+	if (ret<0)
+		luaL_error(L, "event_source_get_enabled failed: %s", strerror(-ret));
+
 	ret = sd_event_source_get_description(evsrc, &desc);
-	lua_pushfstring(L, "event_source [%s] %p", (ret<0)?"unkown":desc, evsrc);
+	lua_pushfstring(L, "event_source [%s, %s] %p", (ret<0)?"unkown":desc, (enabled==-1?"oneshot":(enabled==1?"enabled":"disabled")), evsrc);
```
Comment (azadon) [inline src/evl.c:24]: I am not proud of this, but ... I wanted to keep a oneliner :grimacing:

Comment (kmarkus) [reply src/evl.c:24]: that's fine :+1:

File: src/evl.c [hunk 2]
```diff
@@ -303,22 +303,30 @@ static int timer_callback(sd_event_source *evsrc, uint64_t usec, void* userdata)
...
+		luaL_error(L, "too many arguments");
+	}
+
 	sd_bus *b = lua_checksdbus(L, 1);
 	usec = luaL_checkinteger(L, 2);
 	accuracy = luaL_optinteger(L, 3, 0);
 	luaL_checktype(L, 4, LUA_TFUNCTION);
+	enabled = luaL_optinteger(L, 5, SD_EVENT_ON);
+	if (lua_isnone(L, 5) == 0) {
+		// when there was anything pop it, because stack is used further down.
```
Comment (kmarkus) [inline src/evl.c:327]: Please switch to C-style comments (`/* ... */`)

**Review by kmarkus — APPROVED** [review]
Please update the README with the changed API and add a short entry to the `ChangeLog` section. 

Thank you very much!

**Comment by azadon** [comment]
I've updated README, while doing it I realized that `get_enabled` should return the same type which `set_enabled` accepts. Also, I changed default event mode to ONESHOT, as it makes more sense.

**Comment by kmarkus** [comment]
> I've updated README, while doing it I realized that `get_enabled` should return the same type which `set_enabled` accepts. Also, I changed default event mode to ONESHOT, as it makes more sense.

Thank you for the update. But I don't think we should change the default to ONESHOT, it seems counterintuitive for an API that is called `add_periodic`. Also it will probably break a lot of existing uses.

**Comment by kmarkus** [comment]
> > I've updated README, while doing it I realized that `get_enabled` should return the same type which `set_enabled` accepts. Also, I changed default event mode to ONESHOT, as it makes more sense.
> 
> Thank you for the update. But I don't think we should change the default to ONESHOT, it seems counterintuitive for an API that is called `add_periodic`. Also it will probably break a lot of existing usesA

I think a cleaner way forward would be to introduce a new `_add_time` API that more faithfully covers the main use-cases of SD_EVENT_ADD_TIME(3), including changing expiration time. `add_periodic` could then become a convenience API that uses this internally.

**Pull request merged**